Use only standard Prolog

ekaf opened the pull request:
This PR includes many of the changes from PR #8, aiming to make this library portable across a wider range of Prolog systems.

Additionally, it adds a few improvements: better formatting of  CSV output strings, a predicate to time Prolog calls, and deriving the closure of Wordnet relations.

ekaf committed:
Use only standard Prolog

ekaf committed:
Speed up transitive closure

ekaf committed:
Add ordsets for closure

ekaf commented:
Now, some tests run quicker with gprolog 1.6 than with SWI, so it is questionable whether we should recommend any default system..

ekaf committed:
Improve timeit.pl

ekaf committed:
Address review concerns

ekaf commented on wn_load.pl:
```diff
@@ -1,58 +1,71 @@
-  writef('Loaded %w (%w/%w)\n',[F,P,A]).
+ensure_pred(P):-
+  ( current_predicate(P/A) 
+    -> format(`Already loaded prolog/wn_~w.pl (~w/~w)~n`,[P,P,A])
```
done

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,151 @@
+  apply_call(R,[B,A]),
   glosspair(A,B,G1,G2),
-  writef('Looping %w:\n  from %w (%w)\n    to %w (%w)\n',[R,A,G1,B,G2]),
+  format(`Looping ~w:~n  from ~w (~w)~n    to ~w (~w)`,[R,A,G1,B,G2]),
```
done

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,151 @@
   s(B,N2,W,_,_,_),
   glosspair(A,B,G1,G2),
-  writef('"%w" hyp:\n  from %w-%w (%w)\n    to %w-%w (%w)\n',[W,A,N1,G1,B,N2,G2]),
+  format(`"~w" hyp:~n  from ~w-~w (~w)~n    to ~w-~w (~w)`,[W,A,N1,G1,B,N2,G2]),
```
done

ekaf commented on Makefile:
```diff
@@ -1,6 +1,33 @@
+
+PL ?= swi
+
+# --- fonctions to run Prolog on given file and halt ---
```
done

ekaf commented on utils.pl:
```diff
@@ -0,0 +1,34 @@
+
+ord_add_element([], E, [E]). % Add to empty set
+ord_add_element([H|T], E, [E,H|T]) :-
+    E < H, !. % Insert before element that is larger
```
solved

ekaf commented on wn2csv.pl:
```diff
@@ -1,45 +1,63 @@
+    escape_chars(T, R).
+
+list2csv([A],P):-
+  P=g -> (escape_quotes(A,A1), format(`"~w"~n`,[A1])); format(`~w~n`,[A]).
```
solved

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,38 @@
+    call(R, A, B),
+    \+ ord_memberchk(B, Visited).
+closure(R, A, B, Visited) :-
+    call(R, A, C),
```
solved

ekaf commented on wn2csv.pl:
```diff
@@ -1,45 +1,63 @@
   atom_concat('csv/wn_',P,C1),
   atom_concat(C1,'.csv',C),
-  writef('Writing  %w\n',[C]),
+  format(`Writing ~w~n`,[C]),
```
done

ekaf commented on Makefile:
```diff
@@ -1,6 +1,33 @@
 # wordnet-prolog utilities (c) 2017-25 Eric Kafe
-# SPDX-License-Identifier: Apache-2.0
-# Licensed under the Apache License, Version 2.0
+# License: CC BY 4.0, https://creativecommons.org/licenses/by/4.0/
```
done

ekaf commented on wn_morphy.pl:
```diff
@@ -1,18 +1,17 @@
+:- include(prolog/wn_exc).
+:- include(prolog/wn_s).
 
 % Since v. 7.0, swipl requires this flag for double quotes to produce bytelists:
```
True, but it works with the Prolog systems that we currently support.

ekaf commented on Makefile:
```diff
@@ -1,6 +1,37 @@
+endef
+
+
+# Add others Prolog system here
```
done

ekaf commented on wn_morphy.pl:
```diff
@@ -61,11 +58,16 @@
 
-% Example usage:
-%:-morphy(advertizing, S), writeln(S).
+morph_ini:-.
```
Not sure this suggestion makes any sense

ekaf commented on wn_query.pl:
```diff
@@ -88,25 +86,34 @@
+  consult(wn_load),
+%  load_pred(ant),
+%  time_call(tant),
+%(105974062-1)),
```
done

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,151 @@
 ------------------------------------------ */
 
+% Symmetric relations:
 symrels(['sim','ant','der','vgp']).
```
done

ekaf commented on wn_load.pl:
```diff
@@ -1,58 +1,72 @@
+type_info(Type,Rels):-
+  Term=..[Type,Label,Rels],
+  call(Term),
+  format(`~n~w: ~w~n`, [Label,Rels]).
```
done

ekaf commented on wn_valid.pl:
```diff
@@ -158,8 +177,10 @@
   member(P,LR),
-  pred2arity(P,A,L),
-  writef('Checking duplicates in %w/%w\n',[P,A]),
+  ensure_pred(P), 
```
done

ekaf commented on utils.pl:
```diff
@@ -0,0 +1,42 @@
+    ->  true
+    ;   E @> H  % Continue searching (only if E > H)
+    ->  ord_memberchk(E, T)
+    ).
```
irrelevant

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,44 @@
-% Prevent transitive loops (f. ex. in original WordNet 3.0):
- (A=C, !, writef('Transitive loop: %w %w %w\n', [A,B,C]); true).
+thyp(Start, Hypo) :-
+    closure(hyp, Start, [], Hypo).
```
It is not a bug. The Visited accumulator is intentionally managed as an ordered set to maintain $O(\log N)$ lookup complexity for cycle detection, rather than the $O(N)$ complexity of an unordered list.While the relation (Rel) may produce values in any order, ord_insert/3 is specifically designed to handle this by inserting elements into their correct position in the standard order of terms, ensuring the set remains valid and ordered at every step. Using [] is the standard, portable way to initialize an empty ordered set across ISO-compliant Prologs.

ekaf committed:
Add new review suggestions

pmoura commented on wn_load.pl:
```diff
@@ -1,58 +1,72 @@
+
+type_info(Type,Rels):-
+  Term=..[Type,Label,Rels],
+  call(Term),
```
This is not efficient as it creates a temporary list that later would need to be garbage-collected. As `call/2-6` are ISO Prolog standard built-in predicates, write instead:

```prolog
call(Type, Label, Rels)
```

didoudiaz replied:
Most of these modifications were present in my initial PR proposal.

ekaf replied:
done

pmoura commented on wn_query.pl:
```diff
@@ -51,20 +47,22 @@
   s(I,_,A,_,_,_),
-  apply(R,[I,J]),
+  Term=..[R,I,J],
+  call(Term),
```
This is not efficient as it creates a temporary list that later would need to be garbage-collected. As `call/2-6` are ISO Prolog standard built-in predicates, write instead:

```prolog
call(R, I, J)
```

ekaf replied:
done

pmoura commented on timeit.pl:
```diff
@@ -0,0 +1,53 @@
+  % Convert time to seconds if necessary (e.g., for GNU Prolog, which reports milliseconds).
+  (
+  current_prolog_flag(version_data, Data),
+  Data =.. [gprolog | _]  % Detects GNU Prolog
```
Use instead the de facto standard `dialect` flag.

ekaf replied:
done

didoudiaz commented:
About the `Makefile`: you can avoid the `cleancsv` dep in target `csv` with a `.PHONY: csv` declaration (useful when a target also corresponds to a file/directory):
```
.PHONY: csv

csv:
	@[ -d csv ] || mkdir csv
```
This avoids the `rm -rf csv` followed by a `mkdir csv` (keeping the directory if it exists which is better in particular if a shell has an aldreay executed `cd csv`). NB: you can move up the `.PHONY` delcaration before the first target `all` (as it is usually the case in Makefiles).

Remark: the `cleanpl` target is actually empty (the `rm a.out` is now useless).

pmoura commented on wn_valid.pl:
```diff
@@ -1,132 +1,151 @@
-  apply(R,[A,B]),
-  (apply(R,[B,A]) -> true; writef('Missing %w(%w,%w)\n',[R,B,A])),
+  apply_call(R,[A,B]),
+  (apply_call(R,[B,A]) -> true; format('Missing ~w(~w,~w)~n',[R,B,A])),
```
Calls to the `apply_call/2` predicate where the second argument is a list of fixed length can be replaced with calls to the ISO Prolog standard `call/2-6` built-in predicates. This is the case of 6 out of 8 calls to the `apply_call/2` predicate. Fixing this issue also provides a modest performance improvement.

didoudiaz replied:
Same remark, see my initial PR.

ekaf replied:
done

didoudiaz commented:
Generally speaking, it is better to use a `:- include(File)` compilation directive than executing, at run-time, a `consult(File)` goal. This is particularly true for gprolog if we want to (one day) extend the code for compilation to native code. For this reason I encapsulated all `consult(File)` in a `loadfile(File)` wrapper.

ekaf committed:
Add idempotent file loading and optimize call dispatching

ekaf commented:
Thanks @didoudiaz and @pmoura for the insightful suggestions. I have refactored the project to address your concerns. Specifically:

- Added loader.pl: A lightweight bootstrapper that provides a safe_consult/1 predicate to ensure large files are loaded only once.
- Optimized the dispatcher in utils.pl: Switched from univ ($=..$) to a call/N pattern to reduce overhead.
 
These changes make the scripts more robust and efficient while keeping them self-contained.

ekaf committed:
Replace remaining univ calls

ekaf committed:
Update Readme

ekaf committed:
Fix quotes in CSV output

ekaf commented on wn2csv.pl:
```diff
@@ -1,45 +1,81 @@
 
-:-consult('wn_load.pl').
+:- include(loader).
+
+escape_codes([], []).
+escape_codes([H|T], O) :-
+    (  H = 39          % single quote
+    -> O = [39,39|R]
+    ;  O = [H|R]
+    ),
+    escape_codes(T, R).
+
+escape_quotes(S, Escaped) :-
+  atom_codes(S, Codes),
+  escape_codes(Codes, EscapedCodes),
+  atom_codes(Escaped, EscapedCodes).
```
The handle_index/4 predicate ensures that all strings are quoted.

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,153 @@
-  sk(J,_,K),
-  I\=J.
+  ski(K,I),
+  ski(K,J),
+  I < J.
 
 check_keys:-
-  writeln('Searching for ambiguous sense keys'),
-  findall(X, multikey(X), L),
-  list_to_set(L,S),
-  member(K,S),
-  writef("Ambiguous key: %w\n",[K]),
-  findall(I, sk(I,_,K), IL),
-  list_to_set(IL,IS),
+  write('Searching for ambiguous sense keys'),
+  nl,
+  setof(X, multikey(X), KS),
+  member(K,KS),
+  format('Ambiguous key: ~w~n',[K]),
+  setof(I, ski(K,I), IS),
   member(I,IS),
   g(I,G),
-  writef("    ->%w (%w)\n",[I,G]),
+  format('    ->~w (~w)~n',[I,G]),
   false.
 check_keys:-
-  ok,
-  nl.
+  retractall(ski(_,_)),
+  ok.
```
No, that's not true: the last check_keys clause ensures proper termination.

ekaf committed:
Implement review suggestions

ekaf commented on README.md:
```diff
@@ -51,17 +53,25 @@ avoids duplicates, and contains only unique clauses:
 
 The wordnet-prolog repository also includes alternative branches
-with Prolog versions of WordNet 3.0 and Open English WordNet 2022.
+with Prolog versions of WordNet 3.0 and Open English Wordnet.
```
The two orthographs are used on purpose here. According to WordNet:

- WordNet means Princeton WordNet
- wordnet means any of the machine-readable lexical databases modeled after the Princeton WordNet

This distinction has somewhat eroded, but some authors still tend to maintain it.

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,46 @@
 
-/* ------------------------------------------------------------------
-Synonyms have the same identifier: */
+closure(Rel, Start, Visited, Result) :-
+    % 1. Find an immediate neighbour
+    call(Rel, Start, Next),
+    % 2. Check for cycles using your new utility
+    \+ ord_memberchk(Next, Visited),
+    % 3. Branch: Either this is a result, or we recurse deeper
+    (   Result = Next
+    ;   ord_insert(Visited, Next, NewVisited),
+        closure(Rel, Next, NewVisited, Result)
+    ).
```
qini is run at initialization, so it is not possible to run closure before that.

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,46 @@
 
-/* ------------------------------------------------------------------
-Synonyms have the same identifier: */
+closure(Rel, Start, Visited, Result) :-
+    % 1. Find an immediate neighbour
+    call(Rel, Start, Next),
+    % 2. Check for cycles using your new utility
+    \+ ord_memberchk(Next, Visited),
+    % 3. Branch: Either this is a result, or we recurse deeper
+    (   Result = Next
+    ;   ord_insert(Visited, Next, NewVisited),
+        closure(Rel, Next, NewVisited, Result)
+    ).
```
documented it

ekaf commented on wn_morphy.pl:
```diff
@@ -1,21 +1,20 @@
 % Since v. 7.0, swipl requires this flag for double quotes to produce bytelists:
-:-set_prolog_flag(double_quotes,codes).
+% Please note that some Prolog systems may lack set_prolog_flag/2
+:- set_prolog_flag(double_quotes, codes).
```
done

ekaf commented on loader.pl:
```diff
@@ -0,0 +1,31 @@
+        )
+    ).
+
+:- endif.
```
done

ekaf commented on timeit.pl:
```diff
@@ -0,0 +1,52 @@
+  ; T = T0
+  ).
+
+time_call(Goal) :-
+  % Measure the time taken to execute a given Goal and print the elapsed time.
+  current_time(StartTime),
+  call(Goal),  % Always execute the Goal
+  (
+  StartTime == error
+  -> format('Timing unavailable for ~w.~n', [Goal])
+  ; current_time(EndTime),
+    Elapsed is EndTime - StartTime,
+    format('~w executed in ~2f seconds.~n', [Goal, Elapsed])
+  ).
```
documented it

ekaf commented on Makefile:
```diff
@@ -20,25 +51,23 @@ ps:
 	@echo Deleting CSV files
-	@rm -r csv
+	@rm csv/*.csv
+	@rmdir csv
```
done

ekaf commented on loader.pl:
```diff
@@ -0,0 +1,31 @@
+    (   already_loaded(File)
+    ->  format('~N% Info: ~w already loaded. Skipping.~n', [File])
+    ;   (   format('~N% Consulting: ~w ... ', [File]),
+            flush_output,
```
done

ekaf commented on wn2csv.pl:
```diff
@@ -1,45 +1,81 @@
+  (
+    escape_index(P,N)
+    -> escape_quotes(S, S1),
+       format('''~w''', [S1])
```
done

ekaf commented on wn_load.pl:
```diff
@@ -1,58 +1,63 @@
+
+type_info(Type,Rels):-
+  call(Type,Label,Rels),
+  format(`~n~w: ~w~n`, [Label,Rels]).
```
done

pmoura commented on loader.pl:
```diff
@@ -14,18 +14,21 @@
+         ->  format('~N% Info: ~w already loaded. Skipping.~n', [File])
+         ; ( 
+             format('~N% Consulting: ~w ... ', [File]),
+             ignore(flush_output),
```
The `flush_output/0` predicate is a ISO Prolog Core standard built-in predicate. The Copilot review advice to wrap it with `ignore/1` is pure nonsense. Moreover, `ignore/1` is not even a standard  or de facto standard predicate!

ekaf replied:
Thanks, I have removed _ignore_ again. I find it difficult to get an idea of what the "Iso standard" actually is: the latest document seems to be the 2012 Corrigendum, but it neither mentions _ignore_ nor _flush_output_.

pmoura replied:
The ISO Prolog Core standard corrigenda add/correct only the original 1995 standard. You can get the standard documents from ANSI for a few dozen dollars. The corrigenda can be downloaded freely from the ISO site.

ekaf committed:
Address review comments

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,46 @@
+closure(Rel, Start, Visited, Result) :-
+    % 1. Find an immediate neighbour
+    call(Rel, Start, Next),
+    % 2. Check for cycles using ordered sets 
```
done

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,46 @@
 
-:-loadrels.
+/* ------------------------------------------------------
+Transitive closure of Relation R, starting at Node A
```
According to convention, lower-case "n" is used for wordnets that are not from Princeton.

ekaf commented on wn2csv.pl:
```diff
@@ -1,45 +1,82 @@
+escape_quotes(S, Escaped):-
+  atom_codes(S, Codes),
+  escape_codes(Codes, EscapedCodes),
+  atom_codes(Escaped, EscapedCodes).
```
done

ekaf commented on README.md:
```diff
@@ -51,17 +53,25 @@ avoids duplicates, and contains only unique clauses:
 
 The wordnet-prolog repository also includes alternative branches
-with Prolog versions of WordNet 3.0 and Open English WordNet 2022.
+with Prolog versions of WordNet 3.0 and Open English Wordnet.
```
According to convention, lower-case "n" is used for wordnets that are not from Princeton.

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,153 @@
 ------------------------------------------ */
 
+:- dynamic(ski/2).
+
+mk_ski:-
+  % Make inverse of sk, for prologs that only index first arg
+  for_all(sk(I,_,K), assertz(ski(K,I))).
+
```
done

ekaf commented on loader.pl:
```diff
@@ -0,0 +1,34 @@
+
+----------------------------------------------------------------- */
+
+
+% 1. Declare the tracking predicate as dynamic globally
+:- dynamic(already_loaded/1).
+
+% 2. Define safe_consult/1 only once
+iniloader:-
+  current_predicate(safe_consult/1) -> true
+  ; assertz((
+    safe_consult(File) :-
+      (  already_loaded(File)
+         ->  format('~N% Info: ~w already loaded. Skipping.~n', [File])
+         ; ( 
+             format('~N% Consulting: ~w ... ', [File]),
+             ignore(flush_output),
+             consult(File),
+             assertz(already_loaded(File)),
+             format('Done.~n', [])
+           )
+      )
+    )).
```
This PR is about standard Prolog, so this suggestion is somewhat outside the current priorities. I lean towards merging soon, and consider remaining issues later on.

ekaf commented on wn_query.pl:
```diff
@@ -1,48 +1,46 @@
 
-/* ------------------------------------------------------------------
-Transitive hypernymy and hyponymy: */
+thyp(?Start, ?Hypo)
```
done

ekaf commented:
Thanks @pmoura! Copilot tends to forget things when a conversation goes on for too long. So I am tempted to merge this soon, unless you or @didoudiaz see objections.

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,153 @@
 
 
-Included tests:
+These problems sometimes occurred in past Wordnet versions,
```
see above

ekaf commented on wn_valid.pl:
```diff
@@ -1,132 +1,153 @@
+https://github.com/ekaf/wordnet-prolog/raw/master/wn_valid.pl
 
-SWI-prolog program testing for some potential issues in WordNet
+Tests for a few potential Wordnet database bugs
```
see above

ekaf commented on utils.pl:
```diff
@@ -0,0 +1,45 @@
+dispatch_call(6, P, [A1, A2, A3, A4, A5, A6]) :- call(P, A1, A2, A3, A4, A5, A6).
+
+/* ----------------------------------------------------------------- */
+
```
Done: the predicate is now called forall, since both formulas are logically equivalent (by De Morgan's theorem), despite Copilot's claim that they are distinct. 
However I am not sure that the "half standard" utility predicates are now wrapped in the most adequate way for portability. Every recommended method seems to have limitations, so maybe the _current_predicate_ test could be the least problematic approach.

ekaf committed:
Include the original 1740-bytes Princeton license

ekaf committed:
Define forall

ekaf commented:
This is ready to merge.

The PR achieves its main goal of making the code portable across Prolog systems by replacing SWI-specific constructs with standard/commonly-supported equivalents, while also improving functionality (robust CSV quoting, portable timing) and query behavior/performance (closure-based traversal with cycle protection). The refactors discussed in review (loader bootstrapper / safe single-load, and using call/N rather than =.. for dispatch) have been incorporated.

One item is intentionally deferred because it’s outside the core scope of this PR: adding a mutex/lock around safe_consult/1 (or equivalent) in loader.pl to guard against concurrent loads in multi-threaded runtimes. I’m fine tracking that as a follow-up issue/PR.

Approving.

ekaf commented:
Thanks @didoudiaz and @pmoura!

ekaf commented:
Package released as [WNprolog-3.1.6](https://github.com/ekaf/wordnet-prolog/releases/tag/pwn-3.1.6)
